### ISSUE TITLE
Cake displays raw ANSI output after running specific executables

### ISSUE THREAD

**Issue by nsmithdev** [cake-build/cake#3018]
### What You Are Seeing?
ASCI character are returned to console for any output after running specific executables.
It works fine in the vscode console but exabits the problem in windows cmd console even when using PowerShell.
The same code used to work fine when running the older powershell version of cake but started happening when I switch to the dotnet tool version.
So far I can consistently reproduce with SqlCmd.EXE but it does not happen with something like ping which seems to works fine.
I'm passing an invalid argument for a simple poc that does not require a sql connection.
When the exe is successful for example change `-???` to just `-?` the cake output after the exe is run is still incorrect.

```
Sqlcmd: '??': Unexpected argument. Enter '-?' for help.
←[41m←[37;1mAn error occurred when executing task '←[0m←[41;1m←[37;1mDefault22←[0m←[41m←[37;1m'.←[0m
←[41m←[37;1mError: ←[0m←[41;1m←[37;1mOne or more errors occurred. (Restore database failed)←[0m
←[41m←[37;1m    ←[0m←[41;1m←[37;1mRestore database failed←[0m
```

### What is Expected?
To not see raw ASCI characters 

```
Sqlcmd: '??': Unexpected argument. Enter '-?' for help.
An error occurred when executing task 'Default22'.
Error: One or more errors occurred. (Restore database failed)
    Restore database failed
```

### What version of Cake are you using?
1.0.0-rc0002

### Are you running on a 32 or 64 bit system?
64

### What environment are you running on?  Windows? Linux? Mac?
Window
 
### Are you running on a CI Server?  If so, which one?
No

### How Did You Get This To Happen? (Steps to Reproduce)

```
Task("Default22").Does(() => {
    var settings = new ProcessSettings{ 
      Arguments = @"-???"
    };
    var exitCode = -1;
    using(var process = StartAndReturnProcess(@"SqlCmd.EXE", settings))
    {
        process.WaitForExit();
        exitCode = process.GetExitCode();
    }
    if(exitCode == 1){
        throw new Exception("Restore database failed");
    }
    Information("BlaBla");
    Information("Exit code: {0}", 5);
    Information("Exit code: {0}", exitCode);
    Information("BlaBla");
});
RunTarget("Default22");
```

```
where SqlCmd.EXE
C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\SQLCMD.EXE

dotnet cake .\build.cake
```

**Comment by pascalberger** [comment]
> It works fine in the vscode console but exabits the problem in windows cmd console even when using PowerShell.
The same code used to work fine when running the older powershell version of cake but started happening when I switch to the dotnet tool version.

There's no PowerShell version of Cake, you probably mean [Cake Runner for .NET Framework](https://cakebuild.net/docs/running-builds/runners/cake-runner-for-dotnet-framework). PowerShell is only used for bootstrapper, like you can have a PowerShell bootstrapper for Cake .NET Tool. Beside the host the different runners share most of the codebase including handling of console output.

That being said, what you see was introduced with [Cake 0.38.0](https://cakebuild.net/blog/2020/05/cake-v0.38.0-released) with support for ANSI escape codes (https://github.com/cake-build/cake/issues/2749).

**Comment by nsmithdev** [comment]
I can confirm it was the powershell bootstrapper before.
Downgrading to the 0.37.0 version has restored color and no longer dumps asci after launching the exe.
The only main reason we use windows console is because we double click a .bat file which launches a menu and is where how cake is actually launched.

I attempted setting NO_COLOR env variable but it looks like it didn't make it into version 1.0.0-rc0002 or at least it wasn't working for me.  The NO_COLOR looks like it should work for us when it's released.

So currently by default in windows console there is no color before running the exe and what looks like raw asci after running the command.

![cake](https://user-images.githubusercontent.com/1432565/103469870-29ccf900-4d30-11eb-81f4-b5e4fb905bbf.jpg)

Edition	Windows 10 Pro
Version	20H2
Installed on	‎5/‎28/‎2020
OS build	19042.630
Experience	Windows Feature Experience Pack 120.2212.31.0

Thanks so much for the assistance

**Comment by nsmithdev** [comment]
https://github.com/cake-build/cake/commit/e55cca4375ae3d530ab5d5e0d44237ee503cdec2

**Comment by nsmithdev** [comment]
NO_COLOR env https://github.com/cake-build/cake/commit/e55cca4375ae3d530ab5d5e0d44237ee503cdec2

**Comment by pascalberger** [comment]
> I attempted setting NO_COLOR env variable but it looks like it didn't make it into version 1.0.0-rc0002 or it isn't working for me. The NO_COLOR looks like it should work for us when it's released.

Support for `NO_COLOR` will come in 1.1.

> So currently by default in windows console there is no color before running the exe and what looks like raw asci after running the command.

The ANSI codes oyu see seem to be related to the output of `sqlcmd` as it doesn't happen with other commands and can also be reproduced in PowerShell. Still investigating on this.

**Comment by augustoproiete** [comment]
The issue is that SQLCMD blindly overrides the console mode at the start, effectively disabling ANSI support from that point on. IMO it's a bug of SQLCMD ([I reported it on their UserVoice](https://feedback.azure.com/forums/908035-sql-server/suggestions/42342169-running-sqlcmd-disables-ansi-console-output-suppor), and you might want to [open a ticket with Microsoft](https://support.microsoft.com/hub/4343728/support-for-business) if your company has a paid support plan).

[![image](https://user-images.githubusercontent.com/177608/103492097-1341a300-4dff-11eb-971c-56bbe0280c73.png)](https://docs.microsoft.com/en-us/windows/console/setconsolemode)

From the Cake side, I believe the only thing that can be done is to take a snapshot of the current console mode before starting a process and, after the process ends, restore it back to what it was in case the process changed it.

I opened PR https://github.com/cake-build/cake/pull/3021 that implements this behavior (1) as a possible workaround.

---

(1) With a small caveat: If you decide to write anything to the Console before the process ends, you'll still get raw ANSI output, of course. e.g.

```csharp
using(var process = StartAndReturnProcess(@"SqlCmd.EXE", settings))
{
    // Writing ANSI output here won't work if SqlCmd is still running
    process.WaitForExit();
    // Writing ANSI output from this point onwards is fine as the console mode has been restored
    exitCode = process.GetExitCode();
}
```

**Comment by devlead** [comment]
Fixed by #3021
